File: Android/MeetingManagerAndroid/app/src/main/java/com/microsoft/office365/meetingmgr/HttpHelper.java
/*
 * Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
 * See LICENSE in the project root for license information.
 */
package com.microsoft.office365.meetingmgr;

import android.util.Log;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.office365.meetingmgr.Models.User;

import org.apache.http.HttpResponse;
import org.apache.http.StatusLine;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpRequestBaseHC4;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutionException;

/**
 * Extends functionality of HttpHelperBase by providing app-specific and OData-specific features
 */

@SuppressWarnings("deprecation")
public class HttpHelper extends HttpHelperBase {

    private static boolean mUnified = true;

    private static String mBaseUri;
    private boolean mLogDisabled;
    private String mErrorMessage;

    public HttpHelper() {
        mBaseUri = getOfficeEndPoint() + "Me/";
    }

    public void disableLog(boolean disable) {
        mLogDisabled = disable;
    }

    public String getErrorMessage() {
        return mErrorMessage;
    }

    public <TResult> PagedResult<TResult> getPagedItems(String uri, Class<TResult> clazz) {
        return doGetItems(uri, Integer.MAX_VALUE, 1, clazz);
    }

    public <TResult> List<TResult> getItems(String uri, int count, Class<TResult> clazz) {
        PagedResult<TResult> res = doGetItems(uri, count, Integer.MAX_VALUE, clazz);
        return res.list;
    }

    private <TResult> PagedResult<TResult> doGetItems(String uri, int itemCount, int pageCount, Class<TResult> clazz) {
        String fullUri = buildUri(uri);

        List<TResult> result = new ArrayList<>();
        ObjectMapper mapper = new JsonMapper();

        while (pageCount-- > 0 && itemCount > 0 && fullUri != null) {
            ODataListBase<TResult> list = createList(fullUri);
            list = getItem(fullUri, list.getClass());

            if (list == null) {
                return new PagedResult<>(null, null);
            }

            fullUri = list.getNextLink(fullUri);

            for (Object el : list.value) {
                --itemCount;
                result.add(mapper.convertValue(el, clazz));
            }
        }
        return new PagedResult<>(result, fullUri);
    }

    private <T> ODataListBase<T> createList(String uri) {
        if (HttpHelper.isUnified() || !uri.contains("graph")) {
            return new ODataList<>();
        } else {
            return new ODataListAAD<>();
        }
    }

    private static class ODataList<T> extends ODataListBase<T>  {
        @JsonProperty("@odata.nextLink")
        public String NextLink;

        @Override
        public String getNextLink(String currentUri) {
            return NextLink;
        }
    }

    private static class ODataListAAD<T> extends ODataListBase<T>  {
        @JsonProperty("odata.nextLink")
        public String NextLink;

        @Override
        public String getNextLink(String currentUri) {
            if (Utils.isNullOrEmpty(NextLink)) {
                return null;
            }

            String skiptoken = Utils.getToken(NextLink, "$skiptoken");
            if (skiptoken == null) {
                return null;
            }

            String nextUri = Utils.stripToken(currentUri, "&$skiptoken");
            nextUri = Utils.stripToken(nextUri, "&previous-page");

            if (nextUri.charAt(nextUri.length() - 1) != '&') {
                nextUri += '&';
            }

            nextUri += skiptoken;

            return nextUri;
        }
    }

    private abstract static class ODataListBase<T> {
        abstract String getNextLink(String currentUri);
        public List<T> value;
    }

    public static class PagedResult<TResult> {
        public final List<TResult> list;
        public final String nextUri;

        public PagedResult(List<TResult> list, String nextUri) {
            this.list = list;
            this.nextUri = nextUri;
        }
    }

    @Override
    protected String buildUri(String uri) {
        uri = super.buildUri(uri);

        if (uri.startsWith("http")) {
            return uri;
        }
        return mBaseUri + uri;
    }

    @Override
    protected CloseableHttpResponse executeRequest(HttpRequestBaseHC4 httpRequest) throws ExecutionException, InterruptedException, IOException {
        User user = Manager.Instance.getUser();
        if (user != null) {
            httpRequest.setHeader("AnchorMailbox", user.id);
        }

        setAuthHeader(httpRequest, false);
        CloseableHttpResponse response = super.executeRequest(httpRequest);

        if (needsTokenRefresh(response)) {
            setAuthHeader(httpRequest, true);

            // Repeat failed request with refreshed token
            response = super.executeRequest(httpRequest);
        }

        return response;
    }

    private void setAuthHeader(HttpRequestBaseHC4 httpRequest, boolean refreshToken) {
        String uri = httpRequest.getURI().toString();
        String resourceId = resourceIdFromUri(uri);

        if (resourceId != null) {
            String accessToken = getAccessToken(resourceId, refreshToken);
            httpRequest.setHeader("Authorization", "Bearer " + accessToken);
        }
    }

    private String resourceIdFromUri(String uri) {
        String resourceId = null;

        if (uri.contains("login")) {
            resourceId = null;
        } else if (uri.contains("office")) {
            resourceId = Constants.OFFICE_RESOURCE_ID;
        } else if (uri.contains("graph.windows")) {
            resourceId = Constants.AAD_RESOURCE_ID;
        } else if (uri.contains("graph.microsoft")) {
            resourceId = Constants.OFFICE_RESOURCE_ID_UNIFIED;
        }
        return resourceId;
    }

    private String getAccessToken(String resourceId, boolean isRefresh) {
        return TokenHandler.getAccessToken(resourceId, isRefresh);
    }

    private void writeHttpLog(String message) {
        if (!mLogDisabled) {
            Date dateTime = Calendar.getInstance().getTime();
            String dateTimeString = DateFmt.toFullDateString(dateTime);
            String timedMessage = String.format("%s %s", dateTimeString, message);

            Manager.Instance.getHttpTracer().writeLine(timedMessage);
        }
    }

    @Override
    protected void traceRequest(HttpRequestBaseHC4 httpRequest, String requestBody) {
        String traceMessage = String.format("%s URI=%s", httpRequest.getMethod(), httpRequest.getURI());

        if (requestBody != null) {
            traceMessage = String.format("%s, BODY=%s", traceMessage, requestBody);
        }

        Log.i("HHHH", traceMessage);
        writeHttpLog(traceMessage);
    }

    @Override
    protected void traceResponse(int statusCode, String data) {
        String traceMessage = String.format("%s", statusCode);

        if (!Utils.isNullOrEmpty(data)) {
            if (statusCode < 300) {
                traceMessage += String.format(" RESPONSE=%s", data);
            } else {
                String errMessage = new RestErrorParser().parse(data);
                handleFailure(statusCode, errMessage);

                traceMessage += String.format(" ERROR=%s", errMessage);
            }
        }

        writeHttpLog(traceMessage);
        Log.i("HHHH:RESPONSE=", traceMessage);
    }

    protected void handleFailure(int statusCode, String errMessage) {
        Log.e("HTTP ERROR", String.format("%s", statusCode));
        ErrorLogger.log(String.format("Http Error: %s\n%s", statusCode, errMessage));

        mErrorMessage = errMessage;
    }

    private boolean needsTokenRefresh(HttpResponse response) {
        StatusLine status = response.getStatusLine();

        if (status.getStatusCode() == 401) {
            return true;
        }
        return false;
    }

    public static boolean isUnified() {
        return mUnified;
    }

    public static void setUnified(boolean unified) {
        mUnified = unified;
    }

    private static String getOfficeEndPoint() {
        return mUnified ? Constants.OFFICE_ENDPOINT_UNIFIED : Constants.OFFICE_ENDPOINT;
    }

    public static String getAADEndPoint() {
        return mUnified ? Constants.AAD_ENDPOINT_UNIFIED : Constants.AAD_ENDPOINT;
    }
}

File: Xamarin/Meeting_Manager_Xamarin/Meeting_Manager_Xamarin/Meeting_Manager_Xamarin/HttpHelper.cs
//Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
//See LICENSE in the project root for license information.

using Newtonsoft.Json;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;

namespace Meeting_Manager_Xamarin
{
    class HttpHelper : HttpHelperBase
    {
        private const string BaseGraphUri = @"https://graph.microsoft.com/v1.0/Me/";
        private readonly IAuthenticationService _authService;
        private readonly Logger _logger;

        public HttpHelper(IAuthenticationService authService, Logger logger)
        {
            _authService = authService;
            _logger = logger;
        }

        public async Task<IEnumerable<T>> GetItemsAsync<T>(string uri)
        {
            var list = await GetItemAsync<ODataList<T>>(uri);

            return list == null ? new List<T>() : list.value;
        }

        protected override async Task<TResult> DoHttpAsync<TBody, TResult>(HttpMethod method, string uri, TBody body)
        {
            var request = base.CreateRequest(method, uri, body);
            await SetHeaders(request);

            var requestBody = string.Empty;
            if (request.Content != null)
            {
                requestBody = await request.Content.ReadAsStringAsync();
            }

            var response = await base.ExecuteRequestAsync(request);

            TResult result = default(TResult);
            if (response.IsSuccessStatusCode)
            {
                result = await Deserialize<TResult>(response);
            }
            else
            {
                HandleFailure(await GetErrorMessage(response), response);
            }

            LogResponse(request, requestBody, response);

            return result;
        }

        private async void LogResponse(HttpRequestMessage request, string requestBody, HttpResponseMessage response)
        {
            var method = request.Method.Method;
            var uri = request.RequestUri.ToString();

            var statusCode = $"{(int)response.StatusCode} ({response.StatusCode.ToString()})";
            string responseBody = string.Empty;

            var mediaType = response.Content?.Headers?.ContentType.MediaType;

            if (mediaType != null && !mediaType.Contains("image"))
            {
                responseBody = await response.Content.ReadAsStringAsync();
            }

            _logger.LogHttp(method, uri, requestBody, request.Headers.ToString(),
                        statusCode, responseBody, response.Headers.ToString());
        }

        private async Task SetHeaders(HttpRequestMessage request)
        {
            if (!string.IsNullOrEmpty(App.Me.UserId))
            {
                request.Headers.Add("AnchorMailbox", App.Me.UserId);
            }

            request.Headers.CacheControl = new CacheControlHeaderValue { NoCache = true };  // ???

            await SetAuthorizationHeaderAsync(request);
        }

        protected override string BuildUri(string uri)
        {
            uri = base.BuildUri(uri);

            if (uri.StartsWith("http"))
            {
                return uri;
            }

            return BaseGraphUri + uri;
        }

        protected virtual async void HandleFailure(string errorMessage, HttpResponseMessage response)
        {
            var message = string.IsNullOrEmpty(errorMessage) ?
                    string.Format("Failed with {0}", response.StatusCode) :
                    errorMessage;

            await UI.MessageDialog(message);
        }

        private async Task SetAuthorizationHeaderAsync(HttpRequestMessage request)
        {
            string uri = request.RequestUri.ToString();
            string resourceId = ResourceIdFromUri(uri);

            if (!string.IsNullOrEmpty(resourceId))
            {
                string token = await _authService.GetTokenAsync(resourceId);

                if (!string.IsNullOrEmpty(token))
                {
                    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
                }
            }
        }

        private string ResourceIdFromUri(string uri)
        {
            if (uri.Contains("office"))
            {
                return "https://outlook.office365.com/";
            }
            else if (uri.Contains("graph.microsoft"))
            {
                return "https://graph.microsoft.com/";
            }
            return null;
        }

        public class ODataList<T>
        {
            [JsonProperty("@odata.nextLink")]
            public string NextLink { get; set; }

            public List<T> value { get; set; }
        }
    }
}
